ui/pages/permissions-connect/connect-page/utils.ts

```diff
@@ -69,3 +72,61 @@ export function getCaip25PermissionsResponse(
     },
   };
 }
+
+/**
+ * Gets the default accounts for the requested namespaces.
+ * We need at least one default per requested namespace
+ * if there are more explicitly requested accounts, use those instead
+ * for that namespace
+ *
+ * @param requestedNamespaces - The namespaces requested.
+ * @param supportedRequestedAccounts - The supported requested accounts.
+ * @param allAccounts - All available accounts.
+ */
+export function getDefaultAccounts(
+  requestedNamespaces: CaipNamespace[],
+  supportedRequestedAccounts: MergedInternalAccountWithCaipAccountId[],
+  allAccounts: MergedInternalAccountWithCaipAccountId[],
+): MergedInternalAccountWithCaipAccountId[] {
+  const defaultAccounts: MergedInternalAccountWithCaipAccountId[] = [];
+
+  supportedRequestedAccounts.forEach((account) => {
+    const {
+      chain: { namespace },
+    } = parseCaipAccountId(account.caipAccountId);
+    if (requestedNamespaces.includes(namespace)) {
+      defaultAccounts.push(account);
+    }
+  });
+
+  // sort accounts by lastSelected descending
+  const allAccountsSortedByLastSelected =
+    sortSelectedInternalAccounts(allAccounts);
+
+  for (const namespace of requestedNamespaces) {
+    const defaultAccountExistsForNamespace = defaultAccounts.find((account) => {
```

Comment:
Maybe simpler to keep a list of namespaces that are already satisfied by `supportedRequestedAccounts` so we don't need re-iterate over every default account for every namespace

Reply:
How about this? https://github.com/MetaMask/metamask-extension/pull/29887/commits/1152ec59d329d9479b2b4f117bea2d4ed2dd5a9b